aptos-move/framework/src/natives/string_utils.rs

```diff
@@ -0,0 +1,233 @@
+            write!(out, "{}", u).unwrap();
+        }
+        MoveTypeLayout::Address => {
+            let addr = val.value_as::<move_core_types::account_address::AccountAddress>()?;
+            write!(out, "{}", addr).unwrap();
+        }
+        MoveTypeLayout::Signer => {
+            let signer = val.value_as::<move_core_types::account_address::AccountAddress>()?;
+            write!(out, "{}", signer).unwrap();
+        }
+        MoveTypeLayout::Vector(ty) => {
+            let v = val.value_as::<Vec<Value>>()?;
+            out.push('[');
+            for (i, x) in v.into_iter().enumerate() {
+                if i > 0 {
+                    out.push_str(", ");
+                }
+                native_format_impl(context, ty.as_ref(), x, out)?;
+            }
+            out.push(']');
+        }
+        MoveTypeLayout::Struct(MoveStructLayout::WithTypes { type_, fields, .. }) => {
+            let strct = val.value_as::<Struct>()?;
+            out.push_str(type_.name.as_str());
+            out.push_str(" {");
+            for (i, (ty, x)) in fields.iter().zip(strct.unpack()?).enumerate() {
+                if i > 0 {
+                    out.push_str(", ");
+                }
+                write!(out, "{}: ", ty.name).unwrap();
+                native_format_impl(context, &ty.layout, x, out)?;
+            }
+            out.push('}');
+        }
+        MoveTypeLayout::Struct(MoveStructLayout::WithFields(fields)) => {
+            let strct = val.value_as::<Struct>()?;
+            out.push('{');
+            for (i, (ty, x)) in fields.iter().zip(strct.unpack()?).enumerate() {
+                if i > 0 {
+                    out.push_str(", ");
+                }
+                write!(out, "{}: ", ty.name).unwrap();
+                native_format_impl(context, &ty.layout, x, out)?;
+            }
+            out.push('}');
+        }
+        MoveTypeLayout::Struct(MoveStructLayout::Runtime(fields)) => {
+            let strct = val.value_as::<Struct>()?;
+            out.push('(');
+            for (i, (ty, x)) in fields.iter().zip(strct.unpack()?).enumerate() {
+                if i > 0 {
+                    out.push_str(", ");
+                }
+                native_format_impl(context, ty, x, out)?;
+            }
+            out.push(')');
+        }
+    };
+    Ok(())
+}
+
+pub fn native_format(
+    _gas_params: &FormatGasParams,
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+    ) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert!(ty_args.len() == 1);
+    let ty = context.deref().type_to_fully_annotated_layout(&ty_args[0])?.unwrap();
+    let x = safely_pop_arg!(arguments, Reference);
+    let v = x.read_ref().map_err(|e| SafeNativeError::InvariantViolation(e))?;
+    let mut out = String::new();
+    native_format_impl(context, &ty, v, &mut out)?;
+    let move_str = Value::struct_(Struct::pack(vec![Value::vector_u8(out.into_bytes())]));
```

Comment:
What format does this give?

Reply:
String

Reply:
Lol but in what format? { field_name: field_value } or an array of field values?

Reply:
I added some examples in the comments. But basically "StructName { field1: ... , field2: ... }" for structs and [...,...] for vectors and the rest being trivial

Reply:
I was contemplating 0x1::module::StructName wdyt?

Reply:
0x1::module::StructName makes more sense yeah

Reply:
But it's incredible long and ugly most of the time

Reply:
Up to you then. I don't really see a strong use case for printing structs on chain anyway, may be as some kind of logs stored on chain (e.g. part of events)

Reply:
I added options to either format Foo or 0x1::debug::Foo